### ISSUE TITLE
[Gen2 CF-HPG v1.3 Implementation] Add fine+coarse multiscale readout only

### ISSUE THREAD

**Issue by Irthn1311** [Irthn1311/FER2013_Graph#51]
## Objective

Implement exactly one isolated **CF-HPG v1.3 Multiscale Readout** candidate from accepted v1.2 source, without FER2013 training, validation, Kaggle execution, test access, or follow-up refinement.

This is an **implementation-only** issue.

## Scientific rationale

Accepted single-seed progression:

| Candidate | Clean train acc | Val acc | Acc gap | Registered conclusion |
|---|---:|---:|---:|---|
| v1.0, 64 fine nodes | 0.56505 | 0.52828 | 3.68 pp | underfit |
| v1.1, 144 fine nodes | 0.60539 | 0.57342 | 3.20 pp | resolution partial signal |
| v1.2, deeper raw tokenizer | 0.61489 | 0.58066 | 3.42 pp | tokenizer underfit remains |

v1.0 -> v1.1 produced the materially larger gain; v1.2 added only +0.72 pp validation accuracy and +0.95 pp clean-train accuracy despite a deeper local tokenizer.

The next single hypothesis is therefore **not another tokenizer-depth sweep**.

Current v1.2 performs two fine graph blocks, then compresses `144 x 96` fine nodes through fixed `2x2` arithmetic-mean hierarchy to 36 coarse nodes, and the classifier sees **only the final coarse graph readout**. If fine-stage expression evidence is partially lost by hierarchical averaging/coarse reasoning, the coarse-only readout may be an information bottleneck.

### Hypothesis to isolate

> Retaining a parameter-light global summary of the already-computed fine graph stage and concatenating it with the existing coarse summary should improve representational capacity without changing tokenizer, graph construction, message passing, hierarchy, semantics, or training protocol.

This issue only implements that mechanism. It does not claim the hypothesis is true until a separately preregistered scientific run.

## Exact parent

Branch from exact accepted main commit:

`032c38d49d189000b364d24af7e248eba200acf7`

This is the merge commit containing reviewed v1.2 Tokenizer.

Parent v1.2 model SHA-256:

`36a9a3e9e7b3ca84581a46654b592e26f71ed1cd4a156acd387417c4320986b0`

Parent v1.2 identity:

- parameters `420839`
- trainable variables `66`
- Keras variables `76`

## Exact authorized architecture change

Preserve v1.2 through the end of the fine graph stage:

```text
48x48x1
-> 4x4 raw non-overlapping patches
-> 144 x 16
-> Dense(16->96)
-> GELU
-> Dense(96->96)
-> add positional Dense(2->96)
-> LayerNorm
-> GELU
-> Dropout(0.10)
-> fine hybrid graph
-> exactly 2 Max-Relative blocks, D=96
```

Immediately after the second fine graph block and **before** the existing 2x2 hierarchy, compute:

```text
fine_mean = reduce_mean(fine_nodes, axis=1)   # [B,96]
fine_max  = reduce_max(fine_nodes, axis=1)    # [B,96]
fine_readout = concat([fine_mean, fine_max])  # [B,192]
```

Then continue the v1.2 hierarchy/coarse path **unchanged**:

```text
fine nodes
-> exact parameter-free 2x2 arithmetic mean
-> 36 x 96
-> Dense(96->128)
-> LayerNorm
-> GELU
-> coarse hybrid graph
-> exactly 2 Max-Relative blocks, D=128
```

Existing coarse readout remains:

```text
coarse_mean = reduce_mean(coarse_nodes, axis=1)  # [B,128]
coarse_max  = reduce_max(coarse_nodes, axis=1)   # [B,128]
coarse_readout = concat([coarse_mean, coarse_max]) # [B,256]
```

Replace only the classifier input fusion:

```text
pooled = concat([fine_readout, coarse_readout])  # [B,448]
-> existing readout LayerNorm, now width 448
-> existing readout Dense, now Dense(448->128)
-> GELU
-> Dropout(0.20)
-> Dense(128->7)
```

No learned scale weighting, attention, slots, queries, ROI tokens, anatomical labels, or auxiliary heads.

## Exact expected parameter identity

The only parameter-count change comes from the existing `readout_dense` input width:

- v1.2 `Dense(256->128)`: `32,896` parameters
- v1.3 `Dense(448->128)`: `57,472` parameters
- exact increase: `24,576`

Expected v1.3 total:

`445,415` parameters

Expected trainable-variable count remains:

`66`

Expected total Keras-variable count remains:

`76`

Fail closed if any identity differs.

Target remains comfortably below 0.5M parameters and hard paper budget 0.6M.

## Frozen components — MUST NOT CHANGE

Relative to v1.2, keep all of the following unchanged:

- dataset loader/path guards
- augmentation sequence/probabilities/ranges
- normalization/scaling
- patch size `4x4`
- fine node count `144`
- tokenizer order and widths
- positional projection and additive fusion
- fine graph spatial-8 topology
- cosine top-4 relation graph
- graph UNION/dedup semantics
- graph build-once/reuse-twice lifecycle
- both fine Max-Relative blocks
- exact 2x2 arithmetic-mean hierarchy
- coarse projection `96->128`
- coarse graph construction
- both coarse Max-Relative blocks
- mean/max definition at coarse scale
- classifier hidden width `128`
- logits width `7`
- dropout rates
- optimizer
- LR schedule
- weight decay
- global clipnorm
- batch size
- label smoothing/loss
- checkpoint rule
- early-stopping rule
- seed
- train/validation-only lifecycle
- all test-split isolation guards

`data.py` and `graph.py` should be byte-identical to v1.2 unless an implementation-level import path requires a non-semantic packaging change; any such exception must be explained before review.

## Forbidden changes

Do NOT add or modify:

- CNN/Conv2D/depthwise/separable convolution
- pretrained visual backbone
- MediaPipe
- landmarks
- facial parts / ROI / anatomical masks
- handcrafted HOG/LBP/Gabor/Laplacian features
- attention
- Transformer blocks
- learned slots
- latent/composition queries
- motif banks
- semantic edge labels
- extra graph blocks
- graph top-k
- graph metric
- patch resolution
- tokenizer depth/width
- hierarchy rule
- augmentation/regularization/training policy

Do not implement v1.4.

## Candidate package

Create isolated package, suggested path:

`research/candidates/tf_cf_hpg_v1_3_multiscale_readout/`

Expected files mirror v1.2:

- `__init__.py`
- `data.py`
- `graph.py`
- `model.py`
- `train_validation_only.py`

Do not modify the accepted v1.2 package in place.

## Required focused tests

At minimum prove synthetically:

1. exact patchification stays `[B,144,16]`
2. tokenizer sequence remains byte/structurally equivalent to v1.2
3. fine graph still built once and same adjacency object reused twice
4. fine-stage debug tensor is `[B,144,96]`
5. `fine_readout` is exactly concatenated global mean and global max with shape `[B,192]`
6. hierarchy remains exact non-overlapping arithmetic mean `144->36`
7. coarse graph still built once/reused twice
8. coarse final nodes `[B,36,128]`
9. coarse readout remains mean+max `[B,256]`
10. fused readout is exactly `[fine_mean, fine_max, coarse_mean, coarse_max]` in that order and shape `[B,448]`
11. no learned weighting between scales
12. logits `[B,7]`
13. exact identity `445415 / 66 / 76`
14. difference vs v1.2 is limited to the multiscale readout path/readout width plus isolated packaging/harness references
15. optimizer/scheduler/loss/checkpoint/early-stop config exactly frozen
16. validation path deterministic and augmentation-free
17. test-path isolation remains fail closed
18. no forbidden CNN/anatomy/attention/slot/query/motif paths
19. `tf.function` forward/backward produces finite logits/loss/all gradients
20. AdamW updates parameters
21. save/load round trip reproduces synthetic inference logits
22. frozen Generation-1 / Step-13 checksums remain unchanged

Also run v1.0/v1.1/v1.2/v1.3 focused suites in one process to detect registry/name collisions.

## Implementation evidence

Create an implementation evidence record including:

- exact parent SHA
- branch/head SHA
- changed-file list
- candidate model SHA-256
- exact parameter/variable inventory
- exact v1.2 -> v1.3 architectural diff
- test commands/results
- byte hashes for frozen `data.py` / `graph.py`
- fresh-import result
- PyTorch runtime isolation result
- frozen checksum verifier result
- `git diff --check`

## Scientific boundary

Absolutely no FER2013 train/validation execution under this issue.

Do not launch Kaggle.

Do not access test.

Do not interpret synthetic implementation checks as scientific evidence.

After implementation, open a **Draft PR only** and stop for independent research-lead review.

Final implementation status must be:

`CF_HPG_V1_3_MULTISCALE_READOUT_IMPLEMENTATION_ONLY`

**Comment by Irthn1311** [comment]
## Research-lead amendment — parameter identity correction before any v1.3 implementation outcome

The fail-closed stop is accepted. No scientific attempt was started and no implementation result is being interpreted.

Issue #51 contains an arithmetic omission in the expected parameter identity. The registered architecture itself is unchanged.

For the mandated pipeline

`concat([fine_mean, fine_max, coarse_mean, coarse_max]) [B,448] -> LayerNorm(width=448) -> Dense(448->128) -> GELU -> Dropout -> Dense(128->7)`

the parameter delta relative to accepted v1.2 is:

- `readout_dense`: Dense(256->128) = 32,896 params -> Dense(448->128) = 57,472 params, delta `+24,576`
- `readout_norm`: LayerNorm(256) has 512 trainable params (gamma+beta); LayerNorm(448) has 896, delta `+384`
- exact total delta: `+24,960`

Therefore the authoritative expected v1.3 identity is amended to:

- total parameters: **445,799**
- trainable variables: **66**
- total Keras variables: **76**

No architecture, tokenizer, graph, hierarchy, readout order, training configuration, thresholds, or scientific hypothesis changes are authorized by this amendment. In particular, do not remove, move, freeze, or otherwise alter the existing readout LayerNorm to recover the previously written 445,415 count.

The old `445,415` value and old `+24,576 total increase` statement are superseded only for parameter arithmetic. The registered architecture text remains authoritative.

Implementation may now continue from exact parent `032c38d49d189000b364d24af7e248eba200acf7`, implementation-only, using expected identity `445799 / 66 / 76`.

Still forbidden: FER2013 train/validation execution, Kaggle launch, test access, v1.4, or any unregistered architecture/training change.

**Comment by Irthn1311** [comment]
## Research-lead clarification — v1.3 lifecycle comparator/decision lock before any FER outcome

Architecture implementation in PR #52 is reviewed as structurally correct. One lifecycle ambiguity must be resolved before merge/scientific execution: the isolated v1.3 harness must compare v1.3 against the **accepted v1.2 result**, not reuse v1.2's old v1.1 comparator or `TOKENIZER_*` labels.

No FER2013 v1.3 run has occurred. This clarification is therefore pre-outcome and does not change the registered architecture, optimizer, augmentation, checkpointing, hierarchy, graph, or hypothesis.

### Authoritative v1.2 reference for v1.3 diagnostics

Accepted Issue #50 seed42:

- clean-train accuracy: `0.614894284022432`
- clean-train macro-F1: `0.5614369765915708`
- validation accuracy: `0.5806631373641683`
- validation macro-F1: `0.5238975323290902`

Exact deltas:

- `delta_val_accuracy_pp = 100*(v1.3_val_acc - 0.5806631373641683)`
- `delta_val_macro_pp = 100*(v1.3_val_macro - 0.5238975323290902)`
- `delta_clean_train_accuracy_pp = 100*(v1.3_train_acc - 0.614894284022432)`
- `delta_clean_train_macro_pp = 100*(v1.3_train_macro - 0.5614369765915708)`

### Frozen v1.3 decision precedence

1. `CF_HPG_V1_3_STRETCH_PASS`
2. `CF_HPG_V1_3_PASS`
3. `READOUT_OVERFIT_SHIFT`
4. `READOUT_STRONG_SIGNAL`
5. `READOUT_PARTIAL_SIGNAL`
6. `READOUT_UNDERFIT_REMAINS`
7. `READOUT_INCONCLUSIVE`

Definitions:

- `CF_HPG_V1_3_STRETCH_PASS`: val accuracy >= `0.7000`, val macro-F1 >= `0.6700`, both clean gaps <= `8 pp`.
- `CF_HPG_V1_3_PASS`: val accuracy >= `0.6500`, val macro-F1 >= `0.6200`, both clean gaps <= `8 pp`.
- `READOUT_OVERFIT_SHIFT`: clean-train accuracy improvement vs v1.2 >= `5.0 pp` AND either clean gap > `10 pp`.
- `READOUT_STRONG_SIGNAL`: validation accuracy improvement vs v1.2 >= `5.0 pp` AND clean-train accuracy improvement vs v1.2 >= `5.0 pp`, unless a higher-priority label applies.
- `READOUT_PARTIAL_SIGNAL`: validation accuracy improvement vs v1.2 >= `3.0 pp` and < `5.0 pp`, unless a higher-priority label applies.
- `READOUT_UNDERFIT_REMAINS`: clean-train accuracy < `0.7000` AND validation accuracy < `0.6400`, unless a higher-priority label applies.
- else `READOUT_INCONCLUSIVE`.

These are single-seed practical/diagnostic labels, not statistical claims.

### Required implementation correction in existing PR #52

Patch only the v1.3 lifecycle/tests/evidence so the future harness uses the v1.2 reference and v1.3/readout labels above. Preserve exact decision precedence. Remove stale `V1_1_REFERENCE`, `CF_HPG_V1_2_*`, and `TOKENIZER_*` behavior from the v1.3 package/tests.

Do **not** modify `model.py`, `data.py`, `graph.py`, architecture, parameter identity `445799 / 66 / 76`, optimizer/training config, or test isolation.

Still implementation-only: no FER/Kaggle/test access and no v1.4.
